### PR TITLE
Document HTTP_PROXY and HTTPS_PROXY env vars for agent

### LINKED ISSUE #2853: http_proxy var support
### Description
we should document usage of http_proxy and https_proxy for use cases such as download assets via a proxy or running http checks via a proxy, for nodes which don't have access to the internet

### Category

This is an issue with:

- [ ] errata (error in doc content)
- [ ] new doc (requires a completely new doc)
- [x] update (add missing or refresh existing content)
- [ ] improvement (improve docs content or format)

### Affected pages
Agent reference docs

### PR DESCRIPTION
**Pull request by hillaryfraley**
## Description
Adds section in agent reference to explain how to define `HTTP_PROXY` and `HTTPS_PROXY` environment variables so users can use the defined proxy servers for tasks that require access to the internet (even for unconnected entities).

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/2853

### CODE REVIEW

**Comment by hillaryfraley** [comment]
Review note: I may not have this quite right -- I tried to follow the customer ticket I found that seemed to discuss this issue, but I don't know if I understood correctly. It also seems like this might not be quite enough information to explain how to use these proxy env vars.

**Review by agoddard — COMMENTED** [review]
@cwjohnston @hillaryfraley what do you think about this being generic "if you need to proxy HTTP requests run by the agent (for example downloading assets or executing http checks)" vs making it explicitly about those things?

**Comment by hillaryfraley** [comment]
@agoddard if that makes it more useful or more widely applicable, I think that's better! I may not have taken the best framing approach here.

Would that be something more like this heading instead?
"Use environment variables to define proxies for agent HTTP requests"

I'm only guessing that they are environment variables too (based on the support ticket I read through)

File: content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md [hunk 1]
```diff
@@ -1670,6 +1670,23 @@ This includes your checks and plugins.
 
 For example, if you create a `SENSU_TEST_VAR` variable in your sensu-agent file, it will be available to use in your check configurations as `$SENSU_TEST_VAR`.
 
+##### Use environment variables to define proxies for unconnected entities
```
Comment (cwjohnston) [inline content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md]:
```suggestion
##### Use environment variables to specify an HTTP proxy for agent use
```

File: content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md [hunk 2]
```diff
@@ -1670,6 +1670,23 @@ This includes your checks and plugins.
...
+
+To add dynamic runtime assets and run checks for entities that cannot access the internet, define `SENSU_HTTP_PROXY` and `SENSU_HTTPS_PROXY` environment variables in your sensu-agent file.
```
Comment (cwjohnston) [inline content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md]:
```suggestion
If an HTTP proxy is required to access the internet in your compute environment, you may need to configure Sensu agent for this in order to successfully download dynamic runtime assets or execute commands which depend on internet access.

For Sensu agents which require a proxy server, define `HTTP_PROXY` and `HTTPS_PROXY` environment variables in your sensu-agent file.
```

File: content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md [hunk 3]
```diff
@@ -1670,6 +1670,23 @@ This includes your checks and plugins.
...
+SENSU_HTTPS_PROXY="http://YOUR_PROXY_SERVER:PORT"
+{{< /code >}}
+
+{{% notice note %}}
+**NOTE**: You can use the same proxy server URL for `SENSU_HTTP_PROXY` and `SENSU_HTTPS_PROXY`.
+The proxy server URL you specify for `SENSU_HTTPS_PROXY` does not need to use `https://`.
+{{% /notice %}}
+
+After you add the `SENSU_HTTP_PROXY` and `SENSU_HTTPS_PROXY` environment variables and restart sensu-agent, they will be available to check and hook commands executed by the Sensu agent.
+You can then use `SENSU_HTTP_PROXY` and `SENSU_HTTPS_PROXY` to add dynamic runtime assets, run checks, and complete other tasks that typically require an internet connection for your unconnected entities.
```
Comment (cwjohnston) [inline content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md]:
```suggestion
{{< code shell >}}
HTTP_PROXY="http://YOUR_PROXY_SERVER:PORT"
HTTPS_PROXY="http://YOUR_PROXY_SERVER:PORT"
{{< /code >}}

{{% notice note %}}
**NOTE**: You can use the same proxy server URL for `HTTP_PROXY` and `HTTPS_PROXY`.
The proxy server URL you specify for `HTTPS_PROXY` does not need to use `https://`.
{{% /notice %}}

After you add the `HTTP_PROXY` and `HTTPS_PROXY` environment variables and restart sensu-agent, they will be available to check and hook commands executed by the Sensu agent.
You can then use `HTTP_PROXY` and `HTTPS_PROXY` to add dynamic runtime assets, run checks, and complete other tasks that typically require an internet connection for your unconnected entities.
```

As described in [this stackexchange post](https://unix.stackexchange.com/questions/212894/whats-the-right-format-for-the-http-proxy-environment-variable-caps-or-no-ca), an informal convention has emerged in Unixdom around `HTTP_PROXY` and `HTTPS_PROXY`, as well as the lower-case versions of these variables. 

As I understand, somewhere along the line our implementation uses the [ProxyFromEnvironment](https://golang.org/pkg/net/http/#ProxyFromEnvironment) function in Golang net/http to generate proxy settings based on the values of `HTTP_PROXY` and `HTTPS_PROXY`, but not the lower-case form of same.

Therefore, the environment variables shouldn't be prepended with `SENSU_` in this case.

**Review by cwjohnston — CHANGES_REQUESTED** [review]
Looks like this needs revision for correctness. I added some background and suggestions.

**Comment by cwjohnston** [comment]
> Would that be something more like this heading instead?
> "Use environment variables to define proxies for agent HTTP requests"

I think this makes sense. I see a possibility that folks might conflate "HTTP requests" with the agent websocket connection as websocket can be proxied via HTTP. I don't know for sure but I suspect that the agent websocket connection won't respect these environment variables. We should probably get clarity on this so we can communicate it clearly.

File: content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md (hunk 3 above)
Comment (hillaryfraley) [reply content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md]: Thank you -- I wasn't sure about that but just followed what we had for other env vars.

File: content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md [hunk 4]
```diff
@@ -1670,6 +1670,25 @@ This includes your checks and plugins.
 
 For example, if you create a `SENSU_TEST_VAR` variable in your sensu-agent file, it will be available to use in your check configurations as `$SENSU_TEST_VAR`.
 
+##### Use environment variables to specify an HTTP proxy for agent use
+
+If an HTTP proxy is required to access the internet in your compute environment, you may need to configure Sensu agent for this in order to successfully download dynamic runtime assets or execute commands which depend on internet access.
```
Comment (hillaryfraley) [inline content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md]:
```suggestion
If an HTTP proxy is required to access the internet in your compute environment, you may need to configure the Sensu agent to successfully download dynamic runtime assets or execute commands that depend on internet access.
```

**Comment by hillaryfraley** [comment]
Asked about agent websocket connection and HTTP proxy at https://sensu.slack.com/archives/C60EEQFH8/p1605887768328300

File: content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md [hunk 5]
```diff
@@ -1670,6 +1670,25 @@ This includes your checks and plugins.
 
 For example, if you create a `SENSU_TEST_VAR` variable in your sensu-agent file, it will be available to use in your check configurations as `$SENSU_TEST_VAR`.
 
+##### Use environment variables to specify an HTTP proxy for agent use
+
+If an HTTP proxy is required to access the internet in your compute environment, you may need to configure the Sensu agent to successfully download dynamic runtime assets or execute commands that depend on internet access.
+
+For Sensu agents which require a proxy server, define `HTTP_PROXY` and `HTTPS_PROXY` environment variables in your sensu-agent file.
```
Comment (hillaryfraley) [inline content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md]:
```suggestion
For Sensu agents that require a proxy server, define `HTTP_PROXY` and `HTTPS_PROXY` environment variables in your sensu-agent file.
```

File: content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md [hunk 6]
```diff
@@ -1670,6 +1670,25 @@ This includes your checks and plugins.
...
+{{< /code >}}
+
+{{% notice note %}}
+**NOTE**: You can use the same proxy server URL for `HTTP_PROXY` and `HTTPS_PROXY`.
+The proxy server URL you specify for `HTTPS_PROXY` does not need to use `https://`.
+{{% /notice %}}
+
+After you add the `HTTP_PROXY` and `HTTPS_PROXY` environment variables and restart sensu-agent, they will be available to check and hook commands executed by the Sensu agent.
+You can then use `HTTP_PROXY` and `HTTPS_PROXY` to add dynamic runtime assets, run checks, and complete other tasks that typically require an internet connection for your unconnected entities.
+
```
Comment (hillaryfraley) [inline content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md:1718]:
```suggestion

{{% notice note %}}
**NOTE**: If you define the `HTTP_PROXY` and `HTTPS_PROXY` environment variables, the agent WebSocket connection will also use the proxy URL you specify.
{{% /notice %}}
```

File: content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md [hunk 7]
```diff
@@ -1670,6 +1670,25 @@ This includes your checks and plugins.
...
+{{< code shell >}}
+HTTP_PROXY="http://YOUR_PROXY_SERVER:PORT"
+HTTPS_PROXY="http://YOUR_PROXY_SERVER:PORT"
+{{< /code >}}
+
+{{% notice note %}}
+**NOTE**: You can use the same proxy server URL for `HTTP_PROXY` and `HTTPS_PROXY`.
+The proxy server URL you specify for `HTTPS_PROXY` does not need to use `https://`.
+{{% /notice %}}
```
Comment (hillaryfraley) [inline content/sensu-go/6.1/observability-pipeline/observe-schedule/agent.md]:
```suggestion
You can use the same proxy server URL for `HTTP_PROXY` and `HTTPS_PROXY`.
The proxy server URL you specify for `HTTPS_PROXY` does not need to use `https://`.
```

**Comment by hillaryfraley** [comment]
@cwjohnston I added a note to capture the agent websocket connection info, but I'm not sure I phrased it correctly. Would you take a look?

**Review by cwjohnston — COMMENTED** [review]
Content looks good, but I have not been able to successfully test this functionality in my lab. 

Per https://sensu.slack.com/archives/C60EEQFH8/p1606247300422700 it looks like the functionality is not actually implemented. Please do not merge until we have sorted this out with engineering.

**Comment by hillaryfraley** [comment]
Wondering if the issues here are resolved in https://github.com/sensu/sensu-go/pull/4113 -- that PR only mentions HTTP proxy, not HTTPS.

**Review by cwjohnston — APPROVED** [review]

**Pull request merged**